cloud/migrate.js

```diff
@@ -0,0 +1,319 @@
+  var self = this,
+    _Promise = this._parse.Promise;
+  return function(request, status) {
+    var deadline = new Date() + MAXIMUM_DURATION;
+
+    var lastMigration = _Promise.as(0);
+    console.log("Starting import pass");
+    var totalMigrated = 0;
+    _.each(self._triggers, function(handlers, klass) {
+      if (!def(handlers.migrateObject)) {
+        console.log(klass + " has no migration function; nothing to import");
+        return;
+      } else {
+        console.log("Will import class " +  klass);
+      }
+
+      lastMigration = lastMigration.then(function(migrated) {
+        totalMigrated += migrated;
+        console.log("Starting import of class " + klass);
+        return self._migrateClass(klass, handlers.migrateObject);
+      });
+    });
+    return lastMigration.then(function(migrated) {
+      totalMigrated += migrated;
+    }).then(function() {
+      var message = "Done with an import pass";
+      if (totalMigrated == 0) {
+        message = "Completed initial import!";
+      }
+      console.log(message);
+      status.success(message);
+      return totalMigrated;
+    }, function(error) {
+      return status.error(error)
+    });
+  }
+};
+
+Migrator.prototype._migrateClass = function(klass, migration, deadline) {
+  var self = this,
+    _Promise = this._parse.Promise;
+  if (Date.now() > deadline) {
+    console.log("Shutting down to avoid unclean exit from Parse Cloud Jobs");
+    return _Promise.as(0);
+  }
+
+  // if we use any kind of sort (even objectId) then notEqualTo will eventually
+  // be inefficient and we can time out without even migrating a single object.
+  // We could alternatively keep a separate migration table and track of the most
+  // recent timestamp migrated (using ObjectIDs to break ties). This could work
+  // but tends to make cases like new object migrations much harder. It also
+  // violates the assumption that your hybrid clients can opt out of migrations
+  // by setting "migrationStatus" to 1 on the client.
+  var query = new self._parse.Query(klass)
+    .notContainedIn(
+      consts.MIGRATION_KEY,
+      [consts.IS_MIGRATED, consts.NEEDS_SECOND_PASS, consts.FINISHED_SECOND_PASS]
+    ).limit(consts.BATCH_SIZE);
+
+  // For each batch, map that batch to a migration of a single record and
+  // then setting that record's migation status to done. Then wait for
+  // that batch to complete before resolving the outer promise that lets
+  // us fetch a new batch.
+  return query.find().then(function(objects) {
+    var migrations = _.map(objects, function(object) {
+      return _Promise.as().then(function() {
+        return migration(object);
+      }).then(function() {
+        object.set(consts.MIGRATION_KEY, consts.IS_MIGRATED);
+        return object.save();
+      });
+    });
+    return _Promise.when(migrations).then(function() {
+      return migrations.length;
+    });
+
+  }).then(function(migrated) {
+    // Recursion is the for loop of async.
```

Comment:
I would clarify in the comment that whenever we have migrated BATCH_SIZE objects, that means there might be more still out there.